x-pack/plugin/ml/src/main/java/org/elasticsearch/xpack/ml/action/TransportUpgradeJobModelSnapshotAction.java

```diff
@@ -0,0 +1,258 @@
+                                                  IndexNameExpressionResolver indexNameExpressionResolver,
+                                                  JobConfigProvider jobConfigProvider, MlMemoryTracker memoryTracker,
+                                                  JobResultsProvider jobResultsProvider) {
+        super(UpgradeJobModelSnapshotAction.NAME, transportService, clusterService, threadPool, actionFilters, Request::new,
+            indexNameExpressionResolver, Response::new, ThreadPool.Names.SAME);
+        this.licenseState = licenseState;
+        this.persistentTasksService = persistentTasksService;
+        this.jobConfigProvider = jobConfigProvider;
+        this.jobResultsProvider = jobResultsProvider;
+        this.memoryTracker = memoryTracker;
+        this.migrationEligibilityCheck = new MlConfigMigrationEligibilityCheck(settings, clusterService);
+    }
+
+    @Override
+    protected ClusterBlockException checkBlock(Request request, ClusterState state) {
+        return state.blocks().globalBlockedException(ClusterBlockLevel.METADATA_WRITE);
+    }
+
+    @Override
+    protected void masterOperation(Task task, Request request, ClusterState state,
+                                   ActionListener<Response> listener) {
+        if (migrationEligibilityCheck.jobIsEligibleForMigration(request.getJobId(), state)) {
+            listener.onFailure(ExceptionsHelper.configHasNotBeenMigrated("upgrade job snapshot", request.getJobId()));
+            return;
+        }
+
+        if (licenseState.checkFeature(XPackLicenseState.Feature.MACHINE_LEARNING) == false) {
+            listener.onFailure(LicenseUtils.newComplianceException(XPackField.MACHINE_LEARNING));
+            return;
+        }
+
+        if (state.nodes().getMaxNodeVersion().after(state.nodes().getMinNodeVersion())) {
+            listener.onFailure(ExceptionsHelper.conflictStatusException(
+                "Cannot upgrade job [{}] snapshot [{}] as not all nodes are on version {}. All nodes must be the same version",
+                request.getJobId(),
+                request.getSnapshotId(),
+                state.nodes().getMaxNodeVersion().toString()));
+            return;
+        }
+
+        PersistentTasksCustomMetadata customMetadata = state.getMetadata().custom(PersistentTasksCustomMetadata.TYPE);
+        if (customMetadata != null && (customMetadata.findTasks(
+            MlTasks.JOB_SNAPSHOT_UPGRADE_TASK_NAME,
+            t -> t.getParams() instanceof SnapshotUpgradeTaskParams
+                && ((SnapshotUpgradeTaskParams)t.getParams()).getJobId().equals(request.getJobId())).isEmpty() == false)) {
+            listener.onFailure(ExceptionsHelper.conflictStatusException(
+                "Cannot upgrade job [{}] snapshot [{}] as there is currently a snapshot for this job being upgraded",
+                request.getJobId(),
+                request.getSnapshotId()));
+            return;
```

Comment:
this limitation is effectively a throttling limitation. While conceptually it is possible to upgrade numerous snapshots for the same job, we don't want users to do incidentally overload their cluster and prevent other ML jobs from being started.